apollo-router-core/src/query_planner/mod.rs

```diff
@@ -325,116 +403,68 @@ impl FetchNode {
                     })
                 }
-                Some(Response {
-                    data, mut errors, ..
-                }) => {
-                    // we have to nest conditions and do early returns here
-                    // because we need to take ownership of the inner value
-                    if let Value::Object(mut map) = data {
-                        if let Some(entities) = map.remove("_entities") {
-                            tracing::trace!(
-                                "Received entities: {}",
-                                serde_json::to_string(&entities).unwrap(),
-                            );
-
-                            if let Value::Array(array) = entities {
-                                let mut response = response
-                                    .lock()
-                                    .instrument(tracing::trace_span!("response_lock_wait"))
-                                    .await;
-
-                                let span = tracing::trace_span!("response_insert");
-                                let _guard = span.enter();
-                                for (i, entity) in array.into_iter().enumerate() {
-                                    response.insert_data(
-                                        &current_dir.join(Path::from(i.to_string())),
-                                        entity,
-                                    )?;
-                                }
-
-                                return Ok(());
-                            } else {
-                                return Err(FetchError::ExecutionInvalidContent {
-                                    reason: "Received invalid type for key `_entities`!"
-                                        .to_string(),
-                                });
+            };
+
+            self.response_at_path(current_dir, paths, subgraph_response)
+        }
+
+        #[instrument(, level = "trace", name = "response_insert", skip_all)]
+        fn response_at_path<'a>(
+            &'a self,
+            current_dir: &'a Path,
+            paths: Vec<Path>,
+            subgraph_response: Response,
+        ) -> Result<Value, FetchError> {
+            let Response { data, .. } = subgraph_response;
+
+            if !self.requires.is_empty() {
+                // we have to nest conditions and do early returns here
+                // because we need to take ownership of the inner value
+                if let Value::Object(mut map) = data {
+                    if let Some(entities) = map.remove("_entities") {
+                        tracing::trace!("Received entities: {:?}", &entities);
+
+                        if let Value::Array(array) = entities {
+                            let mut value = Value::default();
+
+                            for (entity, path) in array.into_iter().zip(paths.into_iter()) {
+                                value.insert(&path, entity)?;
```

Comment:
:grey_question: I wonder if we would reduce the number of allocation by making this method takes an owned path :thinking: so you can move the value in it instead of passing it by reference. It gets cloned/copied in there anyway

Reply:
the path is not copied in `insert`, it takes it as reference and iterates over the `PathElement`, so there's no cloning. It's the  `iterate_path` (and `ValueExt::select_values_and_paths`) that allocate paths, and there I went with a reference so a path can be reused in multiple calls, and we do not need to allocate paths for nothing if we do not select any value in the end.
I suspect there will be ways to optimize it further :)

Reply:
Yes sorry I meant insert takes a reference, then in `insert` it's maching over reference, then it either copying (if it's an index, usize) or cloning (key, string, to_string). Therefore, since here you're already dropping the `path` in this loop, you can pass ownership of `path` to `insert` and avoid the cloning of the key string.

Reply:
oh, I see, yes I'll fix that

Reply:
(I might be wrong)

Reply:
ok so if the `Path` is owned, we can avoid the `to_string()` here: https://github.com/apollographql/router/pull/250/files#diff-ee3d76610144cb4b3cfde4d8870c3022107caf2850413af0321a6336ee23c342R262
we would still need to clone the key to have it available here: https://github.com/apollographql/router/pull/250/files#diff-ee3d76610144cb4b3cfde4d8870c3022107caf2850413af0321a6336ee23c342R268
because we need to set the current node before looking up the key again. So there's no gain there :disappointed: